docs/developer-guide/ksqldb-clients/java-client.md

```diff
@@ -0,0 +1,387 @@
+   */
+  CompletableFuture<StreamedQueryResult> streamQuery(String sql);
+  
+  ...
+  
+}
+```
+
+This method may be used to issue both push and pull queries, though the usage pattern is most suited for push queries.
+For pull queries, consider [the `executeQuery()` method](./execute-query.md) instead. 
+
+Query properties can be passed as an optional second argument. See the [client API reference](TODO) for more.
+
+By default, push queries return results starting from the beginning of the stream or table.
+To instead start from the end and only receive newly arriving rows, set the property `auto.offset.reset` to `latest`.
+
+### Asynchronous Usage ###
+
+To consume records in an asynchronous fashion, first create a Reactive Streams subscriber to receive query result rows:
+
+```java
+import io.confluent.ksql.api.client.Row;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+public class RowSubscriber implements Subscriber<Row> {
+
+  private Subscription subscription;
+
+  public RowSubscriber() {
+  }
+
+  @Override
+  public synchronized void onSubscribe(final Subscription subscription) {
+    System.out.println("Subscriber is subscribed.");
+    this.subscription = subscription;
+  }
+
+  @Override
+  public synchronized void onNext(final Row row) {
+    System.out.println("Received a row!");
+    System.out.println("Row: " + row.values());
+  }
+
+  @Override
+  public synchronized void onError(final Throwable t) {
+    System.out.println("Received an error: " + t);
+  }
+
+  @Override
+  public synchronized void onComplete() {
+    System.out.println("Query has ended.");
+  }
+
+  public Subscription getSubscription() {
+    return subscription;
+  }
+}
+```
+
+Then, use the client to send the query result to the server and stream results to the subscriber:
+
+```java
+client.streamQuery("SELECT * FROM MY_STREAM EMIT CHANGES;")
+    .thenAccept(streamedQueryResult -> {
+      System.out.println("Query has started. Query ID: " + streamedQueryResult.queryID());
+      
+      final RowSubscriber subscriber = new RowSubscriber();
+      streamedQueryResult.subscribe(subscriber);
+      subscriber.getSubscription().request(10);
+    }).exceptionally(e -> {
+      System.out.println("Request failed: " + e);
+      return null;
+    });
+```
+
+### Synchronous Usage ###
+
+To consume records one-at-a-time in a synchronous fashion, use the `poll()` method on the query result object.
+If `poll()` is called with no arguments, `poll()` will block until a new row becomes available or the query is terminated.
+You can also pass a `Duration` argument to `poll()` which will cause `poll()` to return null if no new rows are received by the time the duration has elapsed.
+See [the API reference](TODO) for more.
+
+```java
+final StreamedQueryResult streamedQueryResult;
+try {
+  streamedQueryResult = client.streamQuery("SELECT * FROM MY_STREAM EMIT CHANGES;").get();
+} catch (Exception e) {
```

Comment:
I agree it clutters the example but I think it's useful to have a fully fleshed out example that users can copy-paste from. @colinhicks and @MichaelDrogalis mentioned adding to the use case tutorials with examples of how to run tutorial commands from the client instead of the API. Maybe once we do that we can remove the try-catch's here?

Would be curious to hear what @JimGalasyn understands is standard for docs like these -- do we value readability over copy-paste-ability?

Reply:
My 2c: Docs should always be optimised for clarity and readibility and we should avoid anything that detracts from the intention of the example. Otherwise we risk turning off users and giving the impression the API is harder to use than it really is.
Even without try...catch the example is copy and pastable - any Java dev knows they need to handle any checked exceptions.

Reply:
Also let's lose the finals please :)